test/src/test/java/hudson/model/ProjectTest.java

```diff
@@ -175,9 +175,9 @@ public void testGetAssignedLabel() throws Exception{
     public void testGetAssignedLabelString() throws Exception{
         FreeStyleProject p = j.createFreeStyleProject("project");
         Slave slave = j.createOnlineSlave();
-        assertNull("Project should not have any label.", p.getAssignedLabelString());
+        assertEquals("Project should have jenkins master label.", j.jenkins.getSelfLabel().name, p.getAssignedLabelString());
```

Comment:
at first glance this appears wrong - so the message should be clearer.
A project with no node assigned and a Jenkins with **no slaves** should have the jenkins master label.
A project with no node assigned and a Jenkins with nodes should not have a label assigned - as canRoam should be `true`

OTL Now looking at this class there seems to be like a bug in that when you add slaves and set the master to zero executors - the jobs will no longer build and you have to re-save every single one of them - but that is out of scope here.